Properly refresh branches.

rca opened the pull request:
Just calling `git reset --hard` on a branch name doesn't pull in
upstream changes.

rca committed:
Properly refresh branches.

Just calling `git reset --hard` on a branch name doesn't pull in
upstream changes.

s0undt3ch commented on bootstrap-salt.sh:
```diff
@@ -730,7 +730,17 @@ __git_clone_and_checkout() {
+        # if it is a branch name, check out the branch, and pull in the
+        # changes.
+        git branch -a | grep -q ${GIT_REV}
+        if [ "$?" == "0" ]; then
```
``` bash
if [ $? -eq 0 ]; then
```

s0undt3ch commented:
Remember that this is supposed to also run in non bash shells ;)

rca committed:
Updated status test to be cross-shell compatible.

rca commented:
Thanks, update made.

s0undt3ch commented:
No problem, travis would complain about it ;)

[As it did](https://travis-ci.org/saltstack/salt-bootstrap/builds/5350880) ;)

rca committed:
Disable exiting immediately for branch check.

The failing grep will cause the entire script to exit without disabling
immediate exiting.

rca committed:
Moved tag fetching.

In case $GIT_REV is a tag that hasn't been fetched yet.

rca committed:
Reimplemented git reset --hard.

This will make sure the working copy isn't dirty.

rca commented:
I made some other changes that should have fixed the build, but [Travis CI says they are failing](https://travis-ci.org/saltstack/salt-bootstrap/builds/5352164).  I checked the specific test it's complaining about and verified the scenario works and exits with status 0.  Thoughts?

rca commented:
When I run the test suite locally on the `develop` branch I get the same errors that Travis is complaining about, so I don't think it's specifically these changes.  The command I'm running is below, note it's running on the current `develop` head, 60ad2dc9af9d64f07763453a8b0b40513d349c42; can you verify this on your end?

```
local@router01:~$ cd /tmp/salt-bootstrap/tests/
local@router01:/tmp/salt-bootstrap/tests$ git rev-parse HEAD
60ad2dc9af9d64f07763453a8b0b40513d349c42
local@router01:/tmp/salt-bootstrap/tests$ sudo time python runtests.py
[...]
======================================================================
FAIL: test_install_specific_git_sha (bootstrap.test_install.InstallationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/salt-bootstrap/tests/bootstrap/test_install.py", line 364, in test_install_specific_git_sha
    stream_stds=True
  File "/tmp/salt-bootstrap/tests/bootstrap/unittesting.py", line 215, in assert_script_result
    raise AssertionError(err_msg.rstrip())
AssertionError: Failed to install using specific git sha:
 No stdout nor stderr captured. Exit code: 1

======================================================================
FAIL: test_install_specific_git_tag (bootstrap.test_install.InstallationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/salt-bootstrap/tests/bootstrap/test_install.py", line 335, in test_install_specific_git_tag
    stream_stds=True
  File "/tmp/salt-bootstrap/tests/bootstrap/unittesting.py", line 215, in assert_script_result
    raise AssertionError(err_msg.rstrip())
AssertionError: Failed to install using specific git tag:
 No stdout nor stderr captured. Exit code: 1

----------------------------------------------------------------------
Ran 13 tests in 514.093s

FAILED (failures=2, skipped=1)
Command exited with non-zero status 1
322.73user 98.47system 8:37.79elapsed 81%CPU (0avgtext+0avgdata 342944maxresident)k
32936inputs+10374960outputs (427major+6036628minor)pagefaults 0swaps
local@router01:/tmp/salt-bootstrap/tests$
```

Additionally, since the entire suite takes over 5 minutes to run, I've been trying to figure out how to run a single test.  I've tried using the following `unit2` command without success.  Here is what unit2 says to do:

```
local@router01:/tmp/salt-bootstrap/tests$ unit2 -h
[...]
  unit2 test_module.TestClass.test_method - run specified test method
```

But, that doesn't work, any ideas?

```
local@router01:/tmp/salt-bootstrap/tests$ sudo PYTHONPATH=$(pwd) unit2 bootstrap.test_install.InstallationTestCase.test_install_specific_git_tag
Traceback (most recent call last):
  File "/usr/local/bin/unit2", line 9, in <module>
    load_entry_point('unittest2==0.5.1', 'console_scripts', 'unit2')()
  File "/usr/local/lib/python2.7/dist-packages/unittest2/main.py", line 240, in main_
    main(module=None)
  File "/usr/local/lib/python2.7/dist-packages/unittest2/main.py", line 97, in __init__
    self.parseArgs(argv)
  File "/usr/local/lib/python2.7/dist-packages/unittest2/main.py", line 152, in parseArgs
    self.createTests()
  File "/usr/local/lib/python2.7/dist-packages/unittest2/main.py", line 161, in createTests
    self.module)
  File "/usr/local/lib/python2.7/dist-packages/unittest2/loader.py", line 148, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/usr/local/lib/python2.7/dist-packages/unittest2/loader.py", line 133, in loadTestsFromName
    test = obj()
TypeError: unbound method test_install_specific_git_tag() must be called with InstallationTestCase instance as first argument (got nothing instead)
```

s0undt3ch commented:
Well, first of all, I did not add the required code to run as specific test case like on salt, it's on my todo list since I also want that ;)

Second, I had also noticed that the errors were similar if not identical, need to know what else is going on there.

Also, sorry I'm taking so long, but I've been busy trying to fix some stuff on salt, but I'll get to it :)

rca commented:
No problem, thanks!
